File: 138.复制带随机指针的链表.py
#
# @lc app=leetcode.cn id=138 lang=python3
#
# [138] 复制带随机指针的链表
#

# @lc code=start
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""


class Solution:

    def copyRandomList(self, head: 'Node') -> 'Node':
        dict = {}

        root = Node(-1)
        root.next = head
        copy_root = Node(-1)
        cur = copy_root
        while root:
            if root.next:
                if root.next not in dict:
                    dict[root.next] = Node(root.next.val)
                cur.next = dict[root.next]
            if root.random:
                if root.random not in dict:
                    dict[root.random] = Node(root.random.val)
                cur.random = dict[root.random]
            root = root.next
            cur = cur.next
        return copy_root.next


# @lc code=end

File: 138.复制带随机指针的链表.java
/*
 * @lc app=leetcode.cn id=138 lang=java
 *
 * [138] 复制带随机指针的链表
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

import java.util.Map;
import java.util.HashMap;
class Solution {
    public Node copyRandomList(Node head) {
        Map<Node,Node> dict = new HashMap<>();
        Node root = new Node(-1);
        root.next = head;
        Node copyRoot =  new Node(-1);
        Node cur = copyRoot;
        while (root != null) {
            if (root.next != null) {
                if (!dict.containsKey(root.next)) {
                    dict.put(root.next, new Node(root.next.val));
                }
                cur.next = dict.get(root.next);
            }
            if (root.random != null) {
                if (!dict.containsKey(root.random)) {
                    dict.put(root.random, new Node(root.random.val));
                }
                cur.random = dict.get(root.random);
            }
            root = root.next;
            cur = cur.next;
        }
        return copyRoot.next;
        
    }
}
// @lc code=end

